File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
input: [7,4,15,1,10,14,17]

        7
    4       15
  1  10    14  17
  
output: [1, 4, 10, 7, 14, 15, 17]
  
Inorder로 읽었을때 오름차순으로 정렬된다면 BST!

T: O(2N)
S: O(N)
"""
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        output = []
        
        self.inorder(root, output)
        
        for i in range(1, len(output)):
            print(output[i - 1], output[i])
            if output[i - 1] >= output[i]:
                return False
            
        return True
        
    def inorder(self, node, output):
        if node is None:
            return

        self.inorder(node.left, output)
        output.append(node.val)
        self.inorder(node.right, output)

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /*
        Edge Case1:
        root = [5, 4, 6, null, null, 3, 7]
                5
              /  \
             4    6
                 / \
                3   7
    */
    
    TreeNode prev;
    boolean isValid = true;

    public boolean isValidBST(TreeNode root) {
        // Algorithm: Recursion and Inorder
        // Time: O(N), Space: O(N)
        if(root == null) return true;
        
        isValid = isValid && isValidBST(root.left);
        
        if(prev != null && prev.val >= root.val) return false;
        else prev = root;
        
        return isValid && isValidBST(root.right);
    }
}